crates/yttrium/src/sign/mod.rs

```diff
@@ -0,0 +1,1503 @@
+                        ))
+                    })?;
+                let envelope =
+                    envelope_type0::deserialize_envelope_type0(&decoded)
+                        .map_err(|e| PairError::Internal(e.to_string()))?;
+                let key = ChaCha20Poly1305::new(&pairing_uri.sym_key.into());
+                let decrypted = key
+                    .decrypt(&Nonce::from(envelope.iv), envelope.sb.as_slice())
+                    .map_err(|e| PairError::Internal(e.to_string()))?;
+
+                let request =
+                    serde_json::from_slice::<ProposalJsonRpc>(&decrypted)
+                        .map_err(|e| {
+                            PairError::Internal(format!(
+                                "Failed to parse decrypted message: {e}"
+                            ))
+                        })?;
+                if request.method != "wc_sessionPropose" {
+                    return Err(PairError::Internal(format!(
+                        "Expected wc_sessionPropose, got {}",
+                        request.method
+                    )));
+                }
+                tracing::debug!("rpc request: {}", request.id);
+                tracing::debug!(
+                    "{}",
+                    serde_json::to_string_pretty(&request.params).unwrap()
+                );
+                let proposal = request.params;
+                tracing::debug!("{proposal:?}");
+
+                let proposer_public_key = hex::decode(proposal.proposer.public_key)
+                    .map_err(|e| {
+                        PairError::Internal(format!(
+                            "Failed to decode proposer public key: {e}"
+                        ))
+                    })?
+                    .try_into()
+                    .map_err(|_| {
+                        PairError::Internal(
+                            "Failed to convert proposer public key to fixed-size array".to_owned()
+                        )
+                    })?;
+                tracing::debug!("pairing topic: {}", pairing_uri.topic);
+
+                // TODO validate namespaces: https://specs.walletconnect.com/2.0/specs/clients/sign/namespaces#12-proposal-namespaces-must-not-have-chains-empty
+
+                return Ok(SessionProposal {
+                    session_proposal_rpc_id: request.id.into_value(),
+                    pairing_topic: pairing_uri.topic,
+                    pairing_sym_key: pairing_uri.sym_key,
+                    proposer_public_key,
+                    requested_namespaces: proposal
+                        .required_namespaces
+                        .into_iter()
+                        .chain(proposal.optional_namespaces.into_iter())
+                        .collect(),
+                });
+            }
+        }
+
+        Err(PairError::Internal("No message found".to_string()))
+    }
+
+    pub async fn approve(
+        &mut self,
+        pairing: SessionProposal,
```

Comment:
This will be a breaking change for Kotlin, where I pass proposerPubKey which is used to get the SessionProposal from the storage. Sth to discuss

Reply:
What do you mean? I imagine you would lookup the SessionProposal from storage and then pass to this function.

Reply:
> What do you mean? I imagine you would lookup the SessionProposal from storage and then pass to this function.

Yes - thinking the same now. The more question I ask the more I understand how everything should align